accepted/00075-live-patching-enhancements.md

```diff
@@ -0,0 +1,171 @@
+
+1. Add information about the EOL of kernels in the UI
+2. Warn user when a live kernel is approaching EOL, or past EOL
+3. Add CLM filters to make package selection easier
+
+The existing live patching integration in Uyuni is described in the initial [Live Patching RFC](https://github.com/uyuni-project/uyuni-rfc/blob/master/accepted/00027-sle-live-patching.md).
+
+# Detailed design
+[design]: #detailed-design
+
+This design provides solutions for two separate problems regarding to kernel live patching:
+
+1. [Live patching EOL information](#live-patching-eol-information)
+2. [Live patching CLM workflow](#live-patching-clm-workflow)
+
+## 1. Live patching EOL information
+
+SUSE provides lifecycle information on any product to users via 'zypper lifecycle' plugin. The actual data about a product is provided in a plain CSV file in a supplementary package per product. In case of live patching, this package is called `lifecycle-data-sle-module-live-patching` and is installed together with the product. Zypper lifecycle plugin reads this data to display human-friendly information to users.
+
+Uyuni shall use this information to provide the following:
+ - Display EOL dates for the current live patch in system details
+ - Display an out-of-support systems list in the home page
+ - Use the notifications system and/or send emails to report live patched systems approaching/past EOL
+
+### Data Retrieval
+
+This section evaluates three alternative implementations for this feature.
+
+#### Call 'zypper lifecycle' on client
+
+Uyuni calls `zypper lifecycle` on the client and collects the data via Salt during package profile update. The data is then stored in the database per client along with the kernel live version data.
+
+**Pros:**
+ - `zypper lifecycle` is publicly supported
+ - Can be called directly for a package name (a specific live-patch package)
+ - Can be attached to profile update data with an existing process
+
+**Cons:**
+ - Is a 2-step process that requires `zypper lifecycle` to output to a file
+ - `zypper-lifecycle-plugin` needs to be installed on the client
+
+#### Read the lifecycle data file on client
+
+Uyuni runs a Salt state to read the lifecycle data file provided with the package `lifecycle-data-sle-module-live-patching` on the client and collects the data during package profile update. The data is then stored in the database per client along with the kernel live version data.
+
+**Pros:**
+ - Easy to implement
+ - Can be attached to profile update data with an existing process
+
+**Cons:**
+ - File name or format can change in the future
+
+#### Unpack and read the lifecycle data RPM on Uyuni server
+
+The lifecycle information is provided in the `lifecycle-data-sle-module-live-patching` package in the Live Patching product. Uyuni unpacks this RPM and reads the data from the extracted CSV file during reposync of this product. The data is stored in the database per live patch package.
```

Comment:
How complex is such a CSV, can you post an example as snippet?

Reply:
```
cokerunner:~ $> cat /usr/share/lifecycle/data/sle-module-live-patching.lifecycle
# Last updated: 2020-09-22T13:14:02+00:00
# format: name,version,date
# version can contain wildcard character *
kernel-livepatch-4_12_14-23-default,*,2019-06-16
kernel-livepatch-4_12_14-25_13-default,*,2019-08-16
kernel-livepatch-4_12_14-25_16-default,*,2019-08-28
kernel-livepatch-4_12_14-25_19-default,*,2019-10-02
kernel-livepatch-4_12_14-25_22-default,*,2019-10-16
kernel-livepatch-4_12_14-25_25-default,*,2019-10-31
kernel-livepatch-4_12_14-25_3-default,*,2019-07-27
kernel-livepatch-4_12_14-25_6-default,*,2019-08-06
kernel-livepatch-4_12_14-25_28-default,*,2020-02-01
kernel-livepatch-4_12_14-150_14-default,*,2020-03-28
kernel-livepatch-4_12_14-150_17-default,*,2020-05-14
kernel-livepatch-4_12_14-150_22-default,*,2020-06-17
kernel-livepatch-4_12_14-150_27-default,*,2020-07-15
kernel-livepatch-4_12_14-150_32-default,*,2020-08-06
kernel-livepatch-4_12_14-150_35-default,*,2020-09-20
kernel-livepatch-4_12_14-150_38-default,*,2020-10-17
kernel-livepatch-4_12_14-150_41-default,*,2020-11-12
kernel-livepatch-4_12_14-150_47-default,*,2020-12-23
kernel-livepatch-4_12_14-150_52-default,*,2021-06-18
kernel-livepatch-4_12_14-150_55-default,*,2021-08-03
kernel-livepatch-4_12_14-150_58-default,*,2021-09-11
kernel-livepatch-4_12_14-195-default,*,2020-06-24
kernel-livepatch-4_12_14-197_4-default,*,2020-06-24
kernel-livepatch-4_12_14-197_7-default,*,2020-07-04
kernel-livepatch-4_12_14-197_10-default,*,2020-07-15
kernel-livepatch-4_12_14-197_15-default,*,2020-08-06
kernel-livepatch-4_12_14-197_18-default,*,2020-09-20
kernel-livepatch-4_12_14-197_21-default,*,2020-10-18
kernel-livepatch-4_12_14-197_26-default,*,2020-11-12
kernel-livepatch-4_12_14-197_29-default,*,2020-12-17
kernel-livepatch-4_12_14-197_34-default,*,2021-03-02
kernel-livepatch-4_12_14-197_37-default,*,2021-03-31
kernel-livepatch-4_12_14-197_40-default,*,2021-04-29
kernel-livepatch-4_12_14-197_45-default,*,2021-06-10
kernel-livepatch-4_12_14-197_48-default,*,2021-08-03
kernel-livepatch-4_12_14-197_51-default,*,2021-08-07
kernel-livepatch-4_12_14-197_56-default,*,2021-09-09
kernel-livepatch-5_3_18-22-default,*,2021-07-21
kernel-livepatch-5_3_18-24_9-default,*,2021-08-03
kernel-livepatch-5_3_18-24_12-default,*,2021-09-03
kernel-livepatch-5_3_18-24_15-default,*,2021-09-09
```

Reply:
Seems pretty straightforward format wise, thanks

Reply:
and here's one for SLES12:
```
cb-min2:~ # cat /var/lib/lifecycle/data/sle-live-patching.lifecycle
# Last updated: 2020-12-16T16:07:48+00:00
# format: name,version,date
# version can contain wildcard character *
kgraft-patch-3_12_32-33-default,*,2016-01-16
kgraft-patch-3_12_32-33-xen,*,2016-01-16
kgraft-patch-3_12_36-38-default,*,2016-02-10
kgraft-patch-3_12_36-38-xen,*,2016-02-10
kgraft-patch-3_12_38-44-default,*,2016-03-18
kgraft-patch-3_12_38-44-xen,*,2016-03-18
kgraft-patch-3_12_39-47-default,*,2016-04-02
kgraft-patch-3_12_39-47-xen,*,2016-04-02
kgraft-patch-3_12_43-52_6-default,*,2016-06-16
kgraft-patch-3_12_43-52_6-xen,*,2016-06-16
...
```

Reply:
For this option, a high-level algorithm can be:

(at reposync or as a separate task after sync)
```
for all RPMs that start with `lifecycle-data-*`:
  unpack RPM and loop through `*.lifecycle` files:
    read `name,version,date` in file
    if name in [`kgraft-patch-*`, `kernel-livepatch-*`, ...]:
       insert into DB table `susePackageLifecycleData`
    delete temp data
```